### DISCUSSION TITLE
How to address xUnit analyzer warning 1047 properly?

### DISCUSSION THREAD

**Discussion by mu88** [xunit/xunit#2921] [Question]
Dear @bradwilson ,

In several of our projects, the analyzer warning `xUnit1047` (_Avoid using TheoryDataRow arguments that might not be serializable_) recently popped up.

[The docs](https://xunit.net/xunit.analyzers/rules/xUnit1047) recommend to implement the interface `IXunitSerializable`. However, it seems strange to me to clutter a type's contract with a test-platform-specific signature 🤔 but I'm probably missing something, so it would be cool if you could elaborate a bit on the idea behind it.

Thx!

**Comment by bradwilson** [comment]
A very common question is: "Why can't I see my individual data rows in Test Explorer?"

The answer to that is always "you have test data that's not serializable". The purpose of this analyzer (and the other three related rules xUnit1044/xUnit1045/xUnit1046) is to highlight to people when this is the case, so they can stop asking us to figure it out for them. 😄 They are set to INFO level, so they do so non-intrusively.

If this isn't a problem you have/want to solve, you should disable the rule(s).

**Reply by mu88** [reply]
Thx for sharing your ideas, @bradwilson.

However. there probably was a misunderstanding: I'm concerned that I have to decorate my production code with a test-platform-specific attribute. What are your thoughts about that?

**Reply by bradwilson** [reply]
If this is a concern, there are at least a couple options:

- Disable xUnit1047, don't get individually runnable data rows, and go on with your life. To be clear, this is where you already are, except now there's an analyzer to tell you _why_.
- Create test-specific data containers that you can add serialization support to, that internally include your production data types, so that the serialization logic is not on a production data object.

There are more extreme options as well (like constructing the production data types inside the test based on serializable theory data), but those two are the top contenders IMO.
